### PR TITLE
Fix phpstan level 3 errors

### PR DESCRIPTION
**Pull request by blankse**
Partly #6780

Before:
`[ERROR] Found 407 errors`

After:
`[ERROR] Found 324 errors`

### CODE REVIEW

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 1]
```diff
@@ -71,23 +72,21 @@ public function setSubTotalPrice($subTotalPrice)
...
+     * @throws UnsupportedException
      *
-     * @return void
+     * @return float
      */
     public function getSubTotalNetPrice()
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('getSubTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('getSubTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:81]: As this breaks as well, we shouldn't do this in this PR.
I think we should also think about making the method abstract or using interfaces instead.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 2]
```diff
@@ -71,23 +72,21 @@ public function setSubTotalPrice($subTotalPrice)
...
     }
 
     /**
      * @param float $subTotalPrice
      */
     public function setSubTotalNetPrice($subTotalPrice)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setSubTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('setSubTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:89]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 3]
```diff
@@ -111,43 +110,43 @@ public function setTotalPrice($totalPrice)
...
+     * @throws UnsupportedException
      *
-     * @return void
+     * @return float
      */
     public function getTotalNetPrice()
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('getTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('getTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:119]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 4]
```diff
@@ -111,43 +110,43 @@ public function setTotalPrice($totalPrice)
...
     /**
+     * @throws UnsupportedException
+     *
      * @param float $totalPrice
      */
     public function setTotalNetPrice($totalPrice)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('setTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:129]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 5]
```diff
@@ -111,43 +110,43 @@ public function setTotalPrice($totalPrice)
...
-     * Should return an array
+     * @throws UnsupportedException
      *
      * @return array
      */
     public function getTaxInfo()
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('getTaxInfo not implemented for ' . get_class($this));
+        throw new UnsupportedException('getTaxInfo not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:139]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 6]
```diff
@@ -111,43 +110,43 @@ public function setTotalPrice($totalPrice)
...
     /**
+     * @throws UnsupportedException
+     *
      * @param array $taxInfo
      */
     public function setTaxInfo($taxInfo)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setTaxInfo not implemented for ' . get_class($this));
+        throw new UnsupportedException('setTaxInfo not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:149]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php [hunk 7]
```diff
@@ -201,19 +200,19 @@ public function getGiftItems()
...
     /**
+     * @throws UnsupportedException
+     *
      * @param AbstractOrderItem[] $giftItems
      */
     public function setGiftItems($giftItems)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setGiftItems not implemented for ' . get_class($this));
+        throw new UnsupportedException('setGiftItems not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:209]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php [hunk 8]
```diff
@@ -123,43 +124,43 @@ public function setTotalPrice($totalPrice)
...
+     * @throws UnsupportedException
      *
-     * @return void
+     * @return float
      */
     public function getTotalNetPrice()
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('getTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('getTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:133]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php [hunk 9]
```diff
@@ -123,43 +124,43 @@ public function setTotalPrice($totalPrice)
...
     /**
+     * @throws UnsupportedException
+     *
      * @param float $totalNetPrice
      */
     public function setTotalNetPrice($totalNetPrice)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setTotalNetPrice not implemented for ' . get_class($this));
+        throw new UnsupportedException('setTotalNetPrice not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:143]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php [hunk 10]
```diff
@@ -123,43 +124,43 @@ public function setTotalPrice($totalPrice)
...
+     * @throws UnsupportedException
      *
-     * @return void
+     * @return array
      */
     public function getTaxInfo()
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('getTaxInfo not implemented for ' . get_class($this));
+        throw new UnsupportedException('getTaxInfo not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:153]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php [hunk 11]
```diff
@@ -123,43 +124,43 @@ public function setTotalPrice($totalPrice)
...
     /**
+     * @throws UnsupportedException
+     *
      * @param array $taxInfo
      */
     public function setTaxInfo($taxInfo)
     {
-        // @TODO Throw UnsupportedException or change to abstract method in v7.0
-        Logger::err('setTaxInfo not implemented for ' . get_class($this));
+        throw new UnsupportedException('setTaxInfo not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:163]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php [hunk 12]
```diff
@@ -226,6 +227,16 @@ public function setOrderState($orderState)
...
     }
 
+    /**
+     * @throws UnsupportedException
+     *
+     * @param string $comment
+     */
+    public function setComment($comment)
+    {
+        throw new UnsupportedException(__FUNCTION__ . ' is not implemented for ' . get_class($this));
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:237]: Again, I'd revert this change for this PR.

File: bundles/EcommerceFrameworkBundle/Model/ProductInterface.php [hunk 13]
```diff
@@ -19,6 +19,11 @@
  */
 interface ProductInterface
 {
+    /**
+     * @return int
+     */
+    public function getId();
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/Model/ProductInterface.php:25]: Again a proper BC-break, which should be in a dedicated PR. Thanks 😊

File: bundles/EcommerceFrameworkBundle/PriceSystem/ModificatedPriceInterface.php [hunk 14]
```diff
@@ -26,6 +28,11 @@ interface ModificatedPriceInterface extends PriceInterface
      * @return string
      */
     public function getDescription();
+
+    /**
+     * @return RuleInterface|null
+     */
+    public function getRule();
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/PriceSystem/ModificatedPriceInterface.php:35]: Again, I'd revert this change for this PR.

File: composer.json [hunk 15]
```diff
@@ -77,7 +77,7 @@
     "piwik/device-detector": "^3.9",
     "presta/sitemap-bundle": "^1.7 || ^2.1",
     "ramsey/uuid": "^3.8",
-    "sabre/dav": "^3.2",
+    "sabre/dav": "^4.1.1",
```
Comment (brusch) [inline composer.json:80]: As this is out of context, I'd revert this here as well.

File: lib/Event/Model/ElementEventInterface.php [hunk 16]
```diff
@@ -22,4 +22,15 @@ interface ElementEventInterface
...
+    /**
+     * Get argument by key.
+     *
+     * @param string $key Key
+     *
+     * @return mixed Contents of array key
+     *
+     * @throws \InvalidArgumentException If key is not found.
+     */
+    public function getArgument($key);
```
Comment (brusch) [inline lib/Event/Model/ElementEventInterface.php:35]: I'd leave that out for this PR (BC-break).

File: lib/Navigation/Builder.php [hunk 17]
```diff
@@ -166,22 +167,15 @@ public function getNavigation($activeDocument = null, $navigationRootDocument =
...
                     }
-
-                    if ($page instanceof DocumentPage) {
-                        if ($page->getDocumentType() == 'link') {
-                            if ($page->getUri() && strpos($activeDocument->getFullPath(),
-                                    $page->getUri() . '/') === 0) {
-                                $activeTrail = true;
-                            }
-                        }
+                    if (strpos($activeDocument->getFullPath(), $page->getUri() . '/') === 0) {
```
Comment (brusch) [inline lib/Navigation/Builder.php:177]: Where is `if ($page instanceof DocumentPage) {` and `if ($page->getDocumentType() == 'link') {` gone? That's still necessary in my opinion.

File: models/DataObject/ClassDefinition/Data/Localizedfields.php [hunk 18]
```diff
@@ -116,6 +116,16 @@ class Localizedfields extends Data implements CustomResourcePersistingInterface,
      */
     public $fieldDefinitionsCache;
 
+    /**
+     * @var array
+     */
+    public $permissionView = [];
```
Comment (brusch) [inline models/DataObject/ClassDefinition/Data/Localizedfields.php:122]: `permissionView` and `permissionEdit` need to be flushed on sleep.

**Review by brusch — CHANGES_REQUESTED** [review]

File: composer.json (hunk 15 above)
Comment (brusch) [reply composer.json:80]: Just moved that to #7511

Comment (blankse) [reply composer.json:80]: Is because of https://github.com/sabre-io/dav/pull/1269
It will fix an phpstan error. But I can move it out.

Comment (blankse) [reply composer.json:80]: Ok I see, you moved it out :)

Comment (brusch) [reply composer.json:80]: yes, please, I'd love to have this breaking change documented 😉 
Then it's easier for us to track all breaking changes and make a proper upgrade guide 😉

File: bundles/EcommerceFrameworkBundle/CartManager/CartInterface.php [hunk 19]
```diff
@@ -61,6 +61,11 @@ public function isCartReadOnly();
      */
     public function getItem($itemKey);
 
+    /**
+     * @param CartItemInterface[] $items
+     */
+    public function setItems($items);
```
[the remark this answers is not in the document]
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/CartManager/CartInterface.php:67]: https://github.com/pimcore/pimcore/pull/7546

File: lib/Event/Model/ElementEventInterface.php (hunk 16 above)
Comment (blankse) [reply lib/Event/Model/ElementEventInterface.php:35]: https://github.com/pimcore/pimcore/pull/7547

File: bundles/EcommerceFrameworkBundle/Model/ProductInterface.php (hunk 13 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/ProductInterface.php:25]: https://github.com/pimcore/pimcore/pull/7548

File: bundles/EcommerceFrameworkBundle/PriceSystem/ModificatedPriceInterface.php (hunk 14 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/PriceSystem/ModificatedPriceInterface.php:35]: https://github.com/pimcore/pimcore/pull/7549

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 1 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:81]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 2 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:89]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 3 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:119]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 4 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:129]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 5 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:139]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 6 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:149]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php (hunk 7 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrder.php:209]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php (hunk 8 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:133]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php (hunk 9 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:143]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php (hunk 10 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:153]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php (hunk 11 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:163]: https://github.com/pimcore/pimcore/pull/7553

File: bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php (hunk 12 above)
Comment (blankse) [reply bundles/EcommerceFrameworkBundle/Model/AbstractOrderItem.php:237]: https://github.com/pimcore/pimcore/pull/7553

File: lib/Navigation/Builder.php (hunk 17 above)
Comment (blankse) [reply lib/Navigation/Builder.php:177]: @brusch Is in Line 173 (Url is a Subtype of DocumentPage and should always Type 'link', right?), otherwise the getUri method doesn't exist:
`if ($activeDocument instanceof Document && $page instanceof Url && $page->getUri()) {`

Comment (blankse) [reply lib/Navigation/Builder.php]: https://github.com/pimcore/pimcore/pull/7487/commits/1bab47de457d980e81bf71a326862216b6ae506c

File: models/DataObject/ClassDefinition/Data/Localizedfields.php (hunk 18 above)
Comment (blankse) [reply models/DataObject/ClassDefinition/Data/Localizedfields.php:122]: https://github.com/pimcore/pimcore/pull/7487/commits/1b21d73f3c05aec0fe5e7d7b6b55bf6a839f677a

File: lib/Navigation/Builder.php (hunk 17 above)
Comment (brusch) [reply lib/Navigation/Builder.php]: 👌 Thanks!

File: lib/Translation/Translator.php [hunk 20]
```diff
@@ -137,15 +137,21 @@ public function trans($id, array $parameters = [], $domain = null, $locale = nul
...
      * {@inheritdoc}
      */
     public function getLocale()
     {
-        return $this->translator->getLocale();
+        if ($this->translator instanceof LocaleAwareInterface) {
+            return $this->translator->getLocale();
+        }
+
+        return 'en';
```
Comment (brusch) [inline lib/Translation/Translator.php:154]:
```suggestion
        return \Pimcore\Tool::getDefaultLanguage();
```

**Review by brusch — CHANGES_REQUESTED** [review]

**Comment by brusch** [comment]
Nice, thanks a lot!
